File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/analyzers/graphanalysis/utils/TestFileDetector.kt
/**
 * KSA-163: Test file detection heuristics.
 */
package com.codeintel.analyzers.graphanalysis.utils

class TestFileDetector {
    private val testPatterns = listOf(
        Regex("\\.test\\.[jt]sx?$"),
        Regex("\\.spec\\.[jt]sx?$"),
        Regex("_test\\.(py|go|rs)$"),
        Regex("test_.*\\.py$"),
        Regex("Test\\.java$"),
        Regex("Test\\.kt$"),
    )
    private val testDirs = listOf("__tests__", "tests", "test", "spec")

    fun isTestFile(filePath: String): Boolean {
        val normalized = filePath.replace("\\", "/")
        return testPatterns.any { it.containsMatchIn(normalized) } ||
            testDirs.any { "/$it/" in normalized }
    }

    fun isTestFunction(name: String): Boolean =
        name.startsWith("test_") || name.startsWith("Test") ||
            name.startsWith("it_") || name.startsWith("should_") ||
            Regex("^(describe|it|test)\\b").containsMatchIn(name)
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/graph_analysis/utils/test_file_detector.py
"""KSA-163: Test file detection heuristics."""
from __future__ import annotations
import re

_TEST_PATTERNS = [
    re.compile(r"\.test\.[jt]sx?$"), re.compile(r"\.spec\.[jt]sx?$"),
    re.compile(r"_test\.(py|go|rs)$"), re.compile(r"test_.*\.py$"),
    re.compile(r"Test\.java$"), re.compile(r"Test\.kt$"),
]
_TEST_DIRS = ["__tests__", "tests", "test", "spec"]


class TestFileDetector:
    def is_test_file(self, file_path: str) -> bool:
        normalized = file_path.replace("\\", "/")
        return (any(p.search(normalized) for p in _TEST_PATTERNS) or
                any(f"/{d}/" in normalized for d in _TEST_DIRS))

    def is_test_function(self, name: str) -> bool:
        return (name.startswith("test_") or name.startswith("Test") or
                name.startswith("it_") or name.startswith("should_") or
                bool(re.match(r"^(describe|it|test)\b", name)))
